Block and vote handling in node implementation (#454)
nanocurrency/nano-docs: docs/node-implementation/blocks.md

title: Node Implementation - Blocks
description: Review of block handling in the current node implementation

# Node Implementation - Blocks

--8<-- "wip-living-whitepaper.md"

## Block publishing and propagation

> 

Reviewer: ```suggestion
Nodes use a modified gossip protocol for message distribution that enables quick distribution of blocks and votes across the network while distributing the load required to propagate messages across multiple nodes rather than each node having to respond to requests from every other node.  Prioritization of messages are focused on the Principal Representative nodes that make up the core consensus mechanism where they receive blocks and votes directly from other nodes and then messages are spread to the rest of the network via gossiping.

```

> Blocks are initially broadcast and propagated across the network to different type of Representatives and nodes based on the blocks status. Some basic rules are listed below:

Reviewer: ```suggestion
Blocks are initially broadcast and propagated across the network to different types of nodes based on the blocks status. Some basic rules are listed below:
```

- Nodes initially publish new blocks on the live network to **all Principal Representatives they can connect to** and a subset of Non Principal Representative nodes.

> - When a new block is processed on a node that is not a known fork, or is a known fork but the block becomes the new winner on an election, nodes will republish that block to `sqrt(peers)`.

Reviewer: ```suggestion
- When a node processes a new block that is not a known fork, or is a known fork and the block becomes the new winner on an election, nodes will republish that block to `sqrt(peers)`.
```

title: Node Implementation - Blocks
description: Review of block handling in the current node implementation

# Node Implementation - Blocks

--8<-- "wip-living-whitepaper.md"

## Block publishing and propagation

Blocks are initially broadcast and propagated across the network to different type of Representatives and nodes based on the blocks status. Some basic rules are listed below:

> - Nodes initially publish new blocks on the live network to **all Principal Representatives they can connect to** and a subset of Represetnatives which are not Principal Representatives

Reviewer: ```suggestion
- Nodes initially publish new blocks on the live network to **all Principal Representatives they can connect to** and a subset of Non Principal Representatives.
```

> - When a new block is processed on a node that is not a known fork, or is a known fork but the block become the new winner on an election, nodes will republish that block to `sqrt(peers)` (excluding Principal Representatives)

Reviewer: ```suggestion
- When a new block is processed on a node that is not a known fork, or is a known fork but the block become the new winner on an election, nodes will republish that block to `sqrt(peers)`.
```
